### PR TITLE
[BE] Combine OptimizerWrapper and OptimizerContainer

### PR DESCRIPTION
**Pull request by mori360**
Combine `state_dict` and `load_state_dict` from OptimizerWrapper to OptimizerContainer so that we only have one optimzier related class
Also, add `get_lr_scheduler_state` to SchedulersContainer when update `lr_scheduler` at self.state

### CODE REVIEW

File: torchtitan/checkpoint.py [hunk 1]
```diff
@@ -229,35 +228,15 @@ def __init__(
 
         self.states = states
 
+        optimizers.update_for_checkpoint(model_parts)
         self.states.update(
             {
                 "model": ModelWrapper(model_parts),
-                "optimizer": OptimizerWrapper(
```
Comment (tianyu-l) [inline torchtitan/checkpoint.py:235]: we should remove `OptimizerWrapper` since it's no longer used anymore

File: torchtitan/optimizer.py [hunk 2]
```diff
@@ -5,17 +5,26 @@
...
+)
+from torch.distributed.checkpoint.stateful import Stateful
 from torch.optim.lr_scheduler import LambdaLR
 from torchtitan.config_manager import JobConfig
 
 
-class OptimizersContainer:
-    """Util for calling step/zero_grad on multiple optimizers needed for virtual pipeline stages"""
+class OptimizersContainer(Stateful):
+    """Util for calling step/zero_grad on multiple optimizers needed for virtual pipeline stages and save"""
```
Comment (tianyu-l) [inline torchtitan/optimizer.py]: did you mean "and for saving/loading state_dict"?

File: torchtitan/optimizer.py [hunk 3]
```diff
@@ -26,6 +35,14 @@ def __init__(self, model_parts, optimizer_kwargs, name):
                 raise NotImplementedError(f"Optimizer {name} not added.")
             self.optimizers.append(optimizer)
 
+    def update_for_checkpoint(self, model):
```
Comment (tianyu-l) [inline torchtitan/optimizer.py]: why do we need this method and `self.model = []`? can we just do `self.model = model_parts` in `__init__` and other things in save/load_state_dict?

File: torchtitan/optimizer.py [hunk 4]
```diff
@@ -5,17 +5,26 @@
...
 from torch.optim.lr_scheduler import LambdaLR
 from torchtitan.config_manager import JobConfig
 
 
-class OptimizersContainer:
-    """Util for calling step/zero_grad on multiple optimizers needed for virtual pipeline stages"""
+class OptimizersContainer(Stateful):
+    """Util for calling step/zero_grad on multiple optimizers needed for virtual pipeline stages and save"""
 
     def __init__(self, model_parts, optimizer_kwargs, name):
```
Comment (fegin) [inline torchtitan/optimizer.py]: Can we also add typing?

File: torchtitan/optimizer.py [hunk 5]
```diff
@@ -5,17 +5,26 @@
...
         self.optimizers = []
+        self.model = []
```
Comment (fegin) [inline torchtitan/optimizer.py]: It's not clear to me the difference between `model` and `model_parts`. Can we add some comments?

File: torchtitan/optimizer.py [hunk 6]
```diff
@@ -25,22 +38,49 @@ def __init__(self, model_parts, optimizer_kwargs, name):
             else:
                 raise NotImplementedError(f"Optimizer {name} not added.")
             self.optimizers.append(optimizer)
+        self.plain_optim = (
```
Comment (fegin) [inline torchtitan/optimizer.py]: Can we deduplicate `self.plain_optim` and `self.optimizers`? I don't see a reason why do we need to keep both variables.

File: torchtitan/optimizer.py [hunk 7]
```diff
@@ -25,22 +38,49 @@ def __init__(self, model_parts, optimizer_kwargs, name):
...
+
 
 class OptimizersInBackwardContainer(OptimizersContainer):
     """Optimiers in backward to skip .step() and .zero_grad()"""
 
-    def __init__(self, model_parts, optimizer_kwargs, name):
+    def __init__(
+        self, model_parts: List[nn.Module], optimizer_kwargs: Dict[str, Any], name: str
+    ) -> None:
         self.optimizers = []
```
Comment (fegin) [inline torchtitan/optimizer.py:82]: ditto.

File: torchtitan/optimizer.py (hunk 6 above)
Comment (mori360) [reply torchtitan/optimizer.py]: Thank you for the comment. Yeah, `self.plain_optim` is only used for get/load state_dict APIs at checkpoint.
Make `self.optimizers` to be plain this time, from `List[List[optim]]` to `List[optim]` for backward case
Also make some changes related to `optimizers.optimizers`, to remove SchedulersInBackwardContainer and modify assert check at CheckpointManager

File: torchtitan/checkpoint.py [hunk 8]
```diff
@@ -220,44 +176,29 @@ def __init__(
 
             TODO: This is currently unsolved and needs a fix.
         """
-        assert len(model_parts) == len(
+        if job_config.optimizer.early_step_in_backward:
```
Comment (mori360) [inline torchtitan/checkpoint.py]: We make the optimizers.optimizers to be plain for both in backward or not in.
Thus `len(model_parts)` does not work for in backward case.
What do you think of the assert check here and error message?

Comment (tianyu-l) [reply torchtitan/checkpoint.py]: I think we should move the sanity check to `OptimizerContainer`'s `__init__` constructors, rather than doing it here. Basically nothing would've changed after init, and we don't need to check it every time we do checkpoint save.

File: torchtitan/optimizer.py [hunk 9]
```diff
@@ -5,18 +5,31 @@
...
+    and saving/loading optimizer state_dict at checkpoint.
+    """
 
-    def __init__(self, model_parts, optimizer_kwargs, name):
+    def __init__(
+        self, model_parts: List[nn.Module], optimizer_kwargs: Dict[str, Any], name: str
+    ) -> None:
         self.optimizers = []
-        for model in model_parts:
+        self.model = model_parts
```
Comment (tianyu-l) [inline torchtitan/optimizer.py]: I'd prefer calling it `self.model_parts` to be consistent with the plural form of `self.optimizers`

File: torchtitan/optimizer.py [hunk 10]
```diff
@@ -26,21 +39,41 @@ def __init__(self, model_parts, optimizer_kwargs, name):
...
 class OptimizersInBackwardContainer(OptimizersContainer):
     """Optimiers in backward to skip .step() and .zero_grad()"""
 
-    def __init__(self, model_parts, optimizer_kwargs, name):
+    def __init__(
+        self, model_parts: List[nn.Module], optimizer_kwargs: Dict[str, Any], name: str
+    ) -> None:
         self.optimizers = []
-        for model in model_parts:
+        self.model = model_parts
```
Comment (tianyu-l) [inline torchtitan/optimizer.py]: ditto

File: torchtitan/optimizer.py [hunk 11]
```diff
@@ -121,44 +156,30 @@ def linear_warmup_linear_decay(
...
-    def __init__(self, optimizers, lr_lambda):
+    def __init__(self, optimizers, lr_lambda) -> None:
         self.schedulers = []
         for optimizer in optimizers:
             self.schedulers.append(LambdaLR(optimizer, lr_lambda=lr_lambda))
 
-    def step(self):
+    def step(self) -> None:
         for schedulers in self.schedulers:
             schedulers.step()
```
Comment (tianyu-l) [inline torchtitan/optimizer.py]:
```suggestion
        for scheduler in self.schedulers:
            scheduler.step()
```

File: torchtitan/checkpoint.py [hunk 12]
```diff
@@ -220,44 +176,29 @@ def __init__(
...
-                    self.states[f"lr_scheduler_{idx}"] = lr_scheduler[0]
-        else:
-            if len(lr_schedulers.schedulers) == 1:
-                self.states["lr_scheduler"] = lr_schedulers.schedulers[0]
-            else:
-                # For now, pipeline-parallel with looped schedules does not support resharding for lr_scheduler.
-                # It should only support saving and loading a distributed checkpoint with the same number of pp ranks
-                for idx, lr_scheduler in enumerate(lr_schedulers.schedulers):
-                    self.states[f"lr_scheduler_{idx}"] = lr_scheduler
+        self.states.update(lr_schedulers.get_lr_scheduler_state())
```
Comment (tianyu-l) [inline torchtitan/checkpoint.py:188]: Discussed with @mori360 offline: I think as a next step, we should still have a single entry for `lr_scheduler`. To achieve that, we need to understand if any further flattening is needed. This could potentially solve the "PP multi-schedule doesn't support DCP resharding" problem.

File: torchtitan/optimizer.py [hunk 13]
```diff
@@ -121,44 +156,30 @@ def linear_warmup_linear_decay(
...
 
+    def get_lr_scheduler_state(self) -> Dict[str, Any]:
+        state_dict = {}
+        if len(self.schedulers) == 1:
+            state_dict["lr_scheduler"] = self.schedulers[0]
+        else:
+            # For now, pipeline-parallel with looped schedules does not support resharding for lr_scheduler.
+            # It should only support saving and loading a distributed checkpoint with the same number of pp ranks
```
Comment (tianyu-l) [inline torchtitan/optimizer.py:188]: I feel this could be solved, if we do flattening on lr_scheduler state_dict similar to what we did to models and optimizers.

Comment (fegin) [reply torchtitan/optimizer.py:188]: This is hard to achieve. LRScheduler's state_dict is very unstructured. It basically just returns `self.__dict__.items()`, which can be anything. And `LRScheduler` doesn't define a parameter group structure. So different `LRScheduler` may have different implementations. I'm not sure how to flatten `LRScheduler` in a general approach. Unless we only focus on one `LRScheduler`.

Comment (tianyu-l) [reply torchtitan/optimizer.py:188]: @fegin I see. Then maybe we can just focus on the LambdaLR, which should be straightforward -- since every scheduler has the same schedule, we can store only one state and recreate schedulers for each optimizer on the fly when doing checkpoint loading.

Comment (fegin) [reply torchtitan/optimizer.py:188]: ye, I want to emphasize that the direction is correct to flatten LRScheduler, which TorchRec also does, iirc. But because we may not have bandwidths to support all of them, so focusing on one or two and claim what TorchTitan supports is a good idea.

File: torchtitan/optimizer.py [hunk 14]
```diff
@@ -38,6 +38,10 @@ def __init__(
             else:
                 raise NotImplementedError(f"Optimizer {name} not added.")
             self.optimizers.append(optimizer)
+        expected_optim_num = len(self.model_parts)
+        assert expected_optim_num == len(
+            self.optimizers
+        ), "Must pass one optimizer per model part"
```
Comment (tianyu-l) [inline torchtitan/optimizer.py]: how about let's isolate this assertion into a private function `_validate_length(expected_length)` in the base class, so that we can reuse it in any optimizer class

**Review by tianyu-l — APPROVED** [review]
lgtm, thank you! Let's follow up in another PR for the lr scheduler flattening.
Please address final comment before merging.
